src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -950,38 +958,46 @@ private ArrayTypeSymbol VisitArrayInitializer(BoundArrayCreation node)
             var arrayType = (ArrayTypeSymbol)node.Type;
             var elementType = arrayType.ElementType;
 
-            var elementBuilder = ArrayBuilder<BoundExpression>.GetInstance();
-            GetArrayElements(node.InitializerOpt, elementBuilder);
+            var initialization = node.InitializerOpt;
+            var elementBuilder = ArrayBuilder<BoundExpression>.GetInstance(initialization.Initializers.Length);
+            GetArrayElements(initialization, elementBuilder);
 
-            var typeBuilder = (node.Syntax.Kind() == SyntaxKind.ImplicitArrayCreationExpression) ? ArrayBuilder<TypeSymbolWithAnnotations>.GetInstance(elementBuilder.Count) : null;
-            foreach (var element in elementBuilder)
+            int n = elementBuilder.Count;
+            var resultBuilder = ArrayBuilder<Result>.GetInstance(n);
+            for (int i = 0; i < n; i++)
             {
-                VisitRvalue(element);
-                Result elementResult = _result;
-                if (typeBuilder != null)
-                {
-                    typeBuilder.Add(elementResult.Type);
-                }
-                else if (elementType?.IsReferenceType == true)
+                var element = RemoveImplicitConversions(elementBuilder[i]);
```

Comment:
Do we really need to go through all this trouble for an explicitly typed array? Consider splitting implementation for two distinct array initialization modes in order to avoid unnecessary wrapping/unwrapping of conversions, allocating/filling builders with Results, etc.

Reply:
We need to remove and recalculate conversions even in the explicitly typed array case because the conversions in initial binding were calculated ignoring nullability and without the nullability from flow analysis.

We should report a warning on `new A<object>[] { new B<object?>() }` for instance (where `class B<T> : A<T> { }`).

---
In reply to: [169856456](https://github.com/dotnet/roslyn/pull/24935#discussion_r169856456) [](ancestors = 169856456)

Reply:
As you mentioned offline, `VisitConversion` should handle those cases. Added a `// PROTOTYPE` comment for now.